### PR TITLE
EKS: Prompt for MFA token if required

### PR DESCRIPTION
**Pull request by anguslees**
- Enable SDK option to prompt for MFA token refresh if required
  (on stdin)
- Remove hard-coded 30s request timeout, and use inherited context
  instead

### CODE REVIEW

File: kubeprod/pkg/eks/platform.go [hunk 1]
```diff
@@ -115,7 +112,7 @@ func (conf *Config) createHostedZone() (*string, error) {
 // Creates a new user policy (or reuses the existing one) in AWS to allow
 // for integration between External DNS and the corresponding hosted zone
 // in Route 53 zone. The user policy is named like "bbkpr-${dnsZone}".
-func (conf *Config) getUserPolicy() (*string, error) {
+func (conf *Config) getUserPolicy(ctx context.Context) (*string, error) {
```
Comment (falfaro) [inline kubeprod/pkg/eks/platform.go]: Why don't you put the `ctx` variable inside the `conf` structure? This way you save passing it around in functions like `getUserPolicy()`, right?

File: kubeprod/pkg/eks/platform.go [hunk 2]
```diff
@@ -129,15 +126,15 @@ func (conf *Config) getUserPolicy() (*string, error) {
 
 	// Creates (or reuses) the hosted zone in Route 53 to be used for
 	// integration with External DNS
-	hostedZoneID, err := conf.createHostedZone()
+	hostedZoneID, err := conf.createHostedZone(ctx)
```
Comment (falfaro) [inline kubeprod/pkg/eks/platform.go]: Ditto.

File: kubeprod/pkg/eks/platform.go [hunk 3]
```diff
@@ -210,18 +207,18 @@ func (conf *Config) getUserPolicy() (*string, error) {
 
 // Attaches the correct IAM policy to the user used for integration with
 // External DNS.
-func (conf *Config) attachUserPolicy() error {
+func (conf *Config) attachUserPolicy(ctx context.Context) error {
```
Comment (falfaro) [inline kubeprod/pkg/eks/platform.go]: Ditto.

File: kubeprod/pkg/eks/platform.go [hunk 4]
```diff
@@ -238,12 +235,12 @@ func (conf *Config) attachUserPolicy() error {
 // policy attached to it which limits R/W to the hosted Route53 zone
 // to be used by BKPR and R/O for any other zones. The IAM policy
 // will be created if necessary.
-func (conf *Config) createAwsUser() (*string, *string, error) {
+func (conf *Config) createAwsUser(ctx context.Context) (*string, *string, error) {
```
Comment (falfaro) [inline kubeprod/pkg/eks/platform.go]: Ditto.

File: kubeprod/pkg/eks/platform.go [hunk 5]
```diff
@@ -271,7 +268,7 @@ func (conf *Config) createAwsUser() (*string, *string, error) {
 }
 
 // Configuration for integration between External DNS and AWS.
-func (conf *Config) setUpExternalDNS() error {
+func (conf *Config) setUpExternalDNS(ctx context.Context) error {
```
Comment (falfaro) [inline kubeprod/pkg/eks/platform.go]: Ditto.

File: kubeprod/pkg/eks/platform.go [hunk 6]
```diff
@@ -319,7 +316,7 @@ func (conf *Config) describeUserPoolClient(svc *cognitoidentityprovider.CognitoI
 // for integration between OAuth2 Proxy and the AWS Cognito User Pool. The
 // client application is amed like "bbkpr-${dnsZone}" and will be enabled to
 // be used for OpenID Connect.
-func (conf *Config) getUserPoolClient(svc *cognitoidentityprovider.CognitoIdentityProvider, userPoolID string) (*cognitoidentityprovider.UserPoolClientType, error) {
+func (conf *Config) getUserPoolClient(ctx context.Context, svc *cognitoidentityprovider.CognitoIdentityProvider, userPoolID string) (*cognitoidentityprovider.UserPoolClientType, error) {
```
Comment (falfaro) [inline kubeprod/pkg/eks/platform.go]: Ditto.

File: kubeprod/pkg/eks/platform.go [hunk 7]
```diff
@@ -329,13 +326,13 @@ func (conf *Config) getUserPoolClient(svc *cognitoidentityprovider.CognitoIdenti
...
 	for {
-		result, err := svc.ListUserPoolClients(input)
+		result, err := svc.ListUserPoolClientsWithContext(ctx, input)
 		if err != nil {
 			return nil, fmt.Errorf("Error retrieving client applications for user pool ID %s: %v", userPoolID, err)
 		}
 		for _, element := range result.UserPoolClients {
 			if *element.ClientName == clientName {
-				userPoolClient, err := conf.describeUserPoolClient(svc, *element.ClientId, userPoolID)
+				userPoolClient, err := conf.describeUserPoolClient(ctx, svc, *element.ClientId, userPoolID)
```
Comment (falfaro) [inline kubeprod/pkg/eks/platform.go]: Ditto.

File: kubeprod/pkg/eks/platform.go [hunk 8]
```diff
@@ -379,14 +376,14 @@ func (conf *Config) getUserPoolClient(svc *cognitoidentityprovider.CognitoIdenti
...
+func (conf *Config) isValidRegion(region string) bool {
 	rs := endpoints.AwsPartition().Services()[endpoints.CognitoIdpServiceID].Regions()
-	_, ok := rs[conf.OauthProxy.AWSRegion]
+	_, ok := rs[region]
 	return ok
 }
 
 // Configuration for integration between OAuth2 Proxy and AWS Cognito.
-func (conf *Config) setUpOAuth2Proxy() error {
+func (conf *Config) setUpOAuth2Proxy(ctx context.Context) error {
```
Comment (falfaro) [inline kubeprod/pkg/eks/platform.go]: Ditto.

File: kubeprod/pkg/eks/platform.go [hunk 9]
```diff
@@ -395,14 +392,14 @@ func (conf *Config) setUpOAuth2Proxy() error {
...
-			if !conf.isValidRegion() {
+			if !conf.isValidRegion(conf.OauthProxy.AWSRegion) {
 				return fmt.Errorf("AWS region '%s' is not a valid region for the Cognito IDP service", conf.OauthProxy.AWSRegion)
 			}
 		}
 
 		// Configure client ID and client secret required for OAuth2 proxy integration with Cognito
 		svc := cognitoidentityprovider.New(session)
-		userPoolClient, err := conf.getUserPoolClient(svc, conf.OauthProxy.AWSUserPoolID)
+		userPoolClient, err := conf.getUserPoolClient(ctx, svc, conf.OauthProxy.AWSUserPoolID)
```
Comment (falfaro) [inline kubeprod/pkg/eks/platform.go]: Ditto.

File: kubeprod/pkg/eks/platform.go [hunk 10]
```diff
@@ -448,7 +445,7 @@ func (conf *Config) Generate(ctx context.Context) error {
 		//
 		// External DNS setup
 		//
-		err := conf.setUpExternalDNS()
+		err := conf.setUpExternalDNS(ctx)
```
Comment (falfaro) [inline kubeprod/pkg/eks/platform.go]: Ditto.

File: kubeprod/pkg/eks/platform.go [hunk 11]
```diff
@@ -466,7 +463,7 @@ func (conf *Config) Generate(ctx context.Context) error {
 	// oauth2-proxy setup
 	//
 	if conf.OauthProxy.AWSUserPoolID != "" {
-		err := conf.setUpOAuth2Proxy()
+		err := conf.setUpOAuth2Proxy(ctx)
```
Comment (falfaro) [inline kubeprod/pkg/eks/platform.go]: Ditto.

File: kubeprod/pkg/eks/platform.go [hunk 12]
```diff
@@ -50,20 +49,18 @@ func (conf *Config) getAwsSession() *session.Session {
...
+					AssumeRoleTokenProvider: stscreds.StdinTokenProvider,
 				}))
 	}
 	return conf.session
 }
 
 // Retrieves the identity of the caller. Among other details retrieves
 // the AWS account number.
-func (conf *Config) getCallerIdentity() (*sts.GetCallerIdentityOutput, error) {
+func (conf *Config) getCallerIdentity(ctx context.Context) (*sts.GetCallerIdentityOutput, error) {
```
Comment (falfaro) [inline kubeprod/pkg/eks/platform.go]: Why don't you put the `ctx` variable inside the `conf` structure? This way you save passing it around in functions like `getUserPolicy()`, right?

File: kubeprod/pkg/eks/platform.go [hunk 13]
```diff
@@ -73,14 +70,14 @@ func (conf *Config) getCallerIdentity() (*sts.GetCallerIdentityOutput, error) {
 // Creates a new hosted zone in Route 53 if required, or reuses an existing
 // one that matches the fully-qualified name for the DNS zone to be used by
 // BKPR.
-func (conf *Config) createHostedZone() (*string, error) {
+func (conf *Config) createHostedZone(ctx context.Context) (*string, error) {
```
Comment (falfaro) [inline kubeprod/pkg/eks/platform.go]: Ditto.

File: kubeprod/pkg/eks/platform.go [hunk 14]
```diff
@@ -183,7 +180,7 @@ func (conf *Config) getUserPolicy() (*string, error) {
 	if aerr, ok := err.(awserr.Error); ok {
 		if aerr.Code() == iam.ErrCodeEntityAlreadyExistsException {
 			log.Warning("Re-using existing IAM policy for External DNS integration: ", *policyName)
-			callerIdentity, err := conf.getCallerIdentity()
+			callerIdentity, err := conf.getCallerIdentity(ctx)
```
Comment (falfaro) [inline kubeprod/pkg/eks/platform.go]: Ditto.

File: kubeprod/pkg/eks/platform.go [hunk 15]
```diff
@@ -210,18 +207,18 @@ func (conf *Config) getUserPolicy() (*string, error) {
 
 // Attaches the correct IAM policy to the user used for integration with
 // External DNS.
-func (conf *Config) attachUserPolicy() error {
+func (conf *Config) attachUserPolicy(ctx context.Context) error {
 	// Retrieve the ARN for the policy that limits the privileges for
 	// the user to be used for External DNS integration
-	policyArn, err := conf.getUserPolicy()
+	policyArn, err := conf.getUserPolicy(ctx)
```
Comment (falfaro) [inline kubeprod/pkg/eks/platform.go]: Ditto.

File: kubeprod/pkg/eks/platform.go [hunk 16]
```diff
@@ -258,10 +255,10 @@ func (conf *Config) createAwsUser() (*string, *string, error) {
 		log.Infof("Created AKS user: %s", userName)
 	}
 
-	conf.attachUserPolicy()
+	conf.attachUserPolicy(ctx)
```
Comment (falfaro) [inline kubeprod/pkg/eks/platform.go]: Ditto.

File: kubeprod/pkg/eks/platform.go [hunk 17]
```diff
@@ -293,7 +290,7 @@ func (conf *Config) setUpExternalDNS() error {
 	// At this point, if the AWS secret is still empty, try to create an AWS
 	// access key for a user named "bkpr.${dnsZone}"
 	if conf.ExternalDNS.AWSAccessKeyID == "" || conf.ExternalDNS.AWSSecretAccessKey == "" {
-		awsAccessKeyID, awsSecretAccessKey, err := conf.createAwsUser()
+		awsAccessKeyID, awsSecretAccessKey, err := conf.createAwsUser(ctx)
```
Comment (falfaro) [inline kubeprod/pkg/eks/platform.go]: Ditto.

File: kubeprod/pkg/eks/platform.go [hunk 18]
```diff
@@ -304,8 +301,8 @@ func (conf *Config) setUpExternalDNS() error {
 }
 
 // Retrieves information from an existing client application in AWS Cognito
-func (conf *Config) describeUserPoolClient(svc *cognitoidentityprovider.CognitoIdentityProvider, clientID, userPoolID string) (*cognitoidentityprovider.UserPoolClientType, error) {
-	result, err := svc.DescribeUserPoolClient(&cognitoidentityprovider.DescribeUserPoolClientInput{
+func (conf *Config) describeUserPoolClient(ctx context.Context, svc *cognitoidentityprovider.CognitoIdentityProvider, clientID, userPoolID string) (*cognitoidentityprovider.UserPoolClientType, error) {
```
Comment (falfaro) [inline kubeprod/pkg/eks/platform.go]: Ditto.

File: kubeprod/pkg/eks/platform.go (hunk 1 above)
Comment (anguslees) [reply kubeprod/pkg/eks/platform.go]: Passing context as the first arg is the strong golang convention.  The convention comes from the fact that there is no way to state in golang that something should _not_ be shared across threads, other than only ever passing it on the stack (as a function argument).

Comment (anguslees) [reply kubeprod/pkg/eks/platform.go]: See also the intro text from https://golang.org/pkg/context/

Comment (falfaro) [reply kubeprod/pkg/eks/platform.go]: Ah, I thought this was an AWS Context, not a Golang context. Sorry!

**Pull request merged**